Strengthen the precondition for the Bech32 mixed-case mutation test.

jonathanknowles opened the pull request:
# Issue Number

#339 

# Overview

It's possible to change the case of a given character of a Bech32 string and yet not produce a mixed-case string.

In such cases, we can't expect that the `decode` function will fail with a `StringToDecodeHasMixedCase` error.

- [x] To prevent the test failing, I have changed the precondition to actually check that the string is in mixed case.



# Comments

<!-- Additional comments or screenshots to attach if any -->

<!-- 
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Acknowledge any changes required to the Wiki
-->

uroboros commented on src/Cardano/Wallet/Kernel/Read.hs:
```diff
@@ -195,6 +196,13 @@ addressPoolGapByRootId rootId db
+    :: ProtocolMagic
+    -> PublicKey
+    -> Address
+mkEosAddress pm
```
@akegalj this is an oddity that came in with ApplyBlock for EO wallets - what do you think I should do here? Is this ok?

akegalj replied:
So we have moved this function from Cardano.Wallet.Kernel.Wallets to Cardano.Wallet.Kernel.Read . My personal opinion about these changes is not that strong - but `mkEosAddress` looks like a helper function that can reside in both (even in its own Kernel.Address or Kernel.Util module)

I believe it is ok for now

uroboros replied:
👍

denisshevchenko commented:
LGTM, however I'm not deeply in context..

KtorZ commented on src/Cardano/Wallet/Kernel/Pending.hs:
```diff
@@ -73,54 +73,87 @@ newForeign w accountId tx meta = do
+--
+-- Will fail if the HdAccountId does not exist or if some inputs of the
+-- new transaction are not available for spending.
+newTx :: IsNewTxError e
```
This feels wrong, and I now understand why the error below is now being discarded. `e` was here unconstrained to signify that errors could only occurs through the update function, which types is left open for the caller.

Instead, the burden of fetching the state for `IsOurs` should be left to the caller; so instead here, having `IsOurs s => ActiveWallet -> (HdAccountId, s)` would make more sense and remove quite some logic from this function which in the end, doesn't care much if the wallet is external or fully owned; or about the derivation scheme that is being used (hence the abstraction).

Having said that, I know realize something... Why are we even touching this function? We can't create transaction for externally owned wallet through this path; transactions have to be created through the external transactions endpoints. We'll most likely need some rework here when we support the sequential scheme for fully owned wallet but that's completely out-of-scope for E.O. wallets.

uroboros replied:
@KtorZ yes this is required for EO transactions - keep in mind that it does tx submission and updates tx metadata, etc. You'll see that the EO code interacts with this code when `submitSignedTx` hands over to `newPending`. Hence these changes here, which say that based on the account type to which we are adding a tx, we apply the appropriate prefiltering.

uroboros commented on src/Cardano/Wallet/Kernel/Pending.hs:
```diff
@@ -73,54 +73,87 @@ newForeign w accountId tx meta = do
+        prefilterOurs db HdAccountBaseEO{} = do
+            eoAccounts <- liftIO $ getEosPools db (mkEosAddress $ walletPassive ^. walletProtocolMagic)
+            let ours = allOurs eoAccounts
+            return . Right $ (fst <$> ours, coinForThisRoot ours)
```
👍

uroboros commented on src/Cardano/Wallet/Kernel/Pending.hs:
```diff
@@ -73,54 +73,87 @@ newForeign w accountId tx meta = do
+
+    (allAddrs, coinsForRoot) <- ExceptT $ liftIO $
+                                  prefilterOurs db (account ^. hdAccountBase)
+    _ <- liftIO $ upd allAddrs
```
yes, I should return that case statement

rvl committed:
Sqlite: add more locking to DBLayer methods

It seems to be best to put a big lock around all operations which use
the SQLite connection.

I was getting the following failed test:

  λ> quickCheck (prop_parallel db)
  *** Failed! Exception: 'SQLite3 returned ErrorError while attempting to perform step: not an error' (after 5 tests and 3 shrinks):
  ParallelCommands
    { prefix = Commands { unCommands = [] }
    , suffixes =
        [ Pair
            { proj1 =
                Commands
                  { unCommands =
                      [ Command
                          (At (ReadTxHistory (Val (MWid "c"))))
                          (At (Resp (Right (TxHistory (fromList [])))))
                          []
                      ]
                  }
            , proj2 =
                Commands
                  { unCommands =
                      [ Command
                          (At (PutTxHistory (Val (MWid "a")) (fromList [])))
                          (At (Resp (Left (NoSuchWallet (Reference (Symbolic (Var 0)))))))
                          [ Var 0 ]
                      ]
                  }
            }
        ]
    }

jonathanknowles committed:
Strengthen the precondition for the Bech32 mixed-case mutation test.

It's possible to change the case of a given character of a Bech32 string
and yet not produce a mixed-case string.

In such cases, we can't expect that the `decode` function will fail with
a `StringToDecodeHasMixedCase` error.

To prevent the test failing, we change the precondition to actually
check that the string is in mixed case.

KtorZ committed:
Merge pull request #337 from input-output-hk/rvl/154/sqlite-locking

Sqlite: add more locking to DBLayer methods
